Augment details shown by -config with list of applied patches

dra27 opened the pull request:
The main feature of this patch is the addition of an item `patches:` to the output of `ocaml{c,opt} -config`. Also included is a simple facility to allow patches to be included automatically by `configure` or `Makefile.nt` by placing `.patch` files in `patches/` (full details in `patches/Patching.md`). Additionally, two patch files are included which allow this functionality to be applied to distributions 3.09.0&ndash;4.02.3.

The purpose of this is to provide a structured way of indicating back-ported functionality. I have reserved `mpr#n` and `gpr#n` as indicating the back-porting of functionality in a Mantis or GitHub PR but expect distributions (i.e. OPAM) to come up with further conventions as required. The idea is that, for example, if one is trying to compile, say, dose3 on Windows where the fixes in MPR#6120 are required that a configure script can either determine that the OCaml version is >= 4.03.0 _or_ that `patches:` is found in `ocaml{c,opt} -config` and includes `mpr#6120`.

dra27 committed:
Display applied patches as part of -config option

Augment the information displayed by the -config option of ocamlc and
ocamlopt to include details of installed patches. utils/config.ml now
contains a string list patches whose elements are the filenames (without
extension) of patches/*.patched (thus, if a patch is applied, it need
simply touch patches/foo.patched to have foo included in the list of
applied patches).

Additionally, if a patch file is placed in patches/
(e.g. patches/foo.patch) then configure or Makefile.nt will apply
foo.patch (and automatically touch foo.patched) and
make [-f Makefile.nt] distclean will automatically reverse any patches
applied this way.

The intention is to provide a slightly more fine-grained approach than
encoding everything in +add-on in the OCaml version number.

dra27 committed:
Provide patches for back-porting GPR#433

The -config option for ocamlc and ocamlopt was introduced in
OCaml 3.09.0. Patches are provided which allow 3.09.0-4.03.0 to be
augmented with the patching system provided in OCaml 4.04.0.

dra27 commented:
If this is accepted, I'm hoping that there are a couple of places where it could quickly/helpfully be integrated:
- opam-repository for compiler switches (hello @avsm, @altgr!), so that all the previous switches at least include the "patches: " item (it's often useful to back-port fixes to the Unix library for the Windows distros, and OPAM will obviously be a fantastic way of managing this - I also longer term have a skeleton idea for patch-as-a-package - i.e. that, say, ocaml-pr6120 becomes a package that can be installed to an older version)
- OCPWin (hello @lefessan!) for example, to indicate back-ported compiler features in the 4.01 distribution (e.g. `patches: 4.02/match-exception`)

damiendoligez commented on .gitattributes:
```diff
@@ -18,6 +18,8 @@ CONTRIBUTING.md          ocaml-typo=missing-header
 INSTALL                  ocaml-typo=missing-header
 LICENSE                  ocaml-typo=non-printing,missing-header
 
+Makefile.shared          ocaml-typo=long-line
```
Don't do that. We have very good reasons to limit lines to 80 characters and exceptions should be rare.

dra27 replied:
OK - out of curiosity only, what are they? I found years ago working with `MicrosoftLongNamedAPI(VERY_LONG_NAMED_CONSTANT,)` resulted in 80 characters vastly reducing readability in too many places and so switched to 120-150 (especially as screens got wider, rather than longer)

dra27 replied:
I've changed my added lines in Makefile.shared to be less than 81 chars, but L20 in Makefile.shared was already longer (which might be why I updated in .gitattributes in the first place!)

damiendoligez commented on .gitattributes:
```diff
@@ -39,6 +41,8 @@ ocamlbuild/TODO          ocaml-typo=missing-header
 ocamldoc/Changes.txt     ocaml-typo=missing-header
 ocamldoc/ocamldoc.sty    ocaml-typo=missing-header
 
+patches/Patches.md       ocaml-typo=very-long-line,long-line,missing-header
```
Same here, do you have any reason not to cut the lines in a `.md` file?

dra27 replied:
Line breaks are respected in Markdown (at least that seemed to be what was happening?) - if there's a way of escaping them (a la an end-of-line `%` in TeX) then great, but formatted with 80 character lines the rendered result had some very confusing line breaks.

dbuenzli replied:
> Line breaks are respected in Markdown (at least that seemed to be what was happening?)

No, markdown follows the usual paragraphs are delimited by an empty line. Line breaks are only respected if you end a line with two spaces.

dra27 replied:
Thanks - it's a feature/bug of GitHub MarkDown it appears! My MarkDown editor uses the GitHub API for rendering, which appears to "respect" all the line breaks in the file, though its internal MarkDown renderer works correctly.
I've updated it to use 80 characters, therefore.

damiendoligez commented:
This is interesting but a significant change, and I expect a rather long discussion. As 4.03 is already very late, this will be for 4.04.

dra27 commented:
It would appear trunk is broken at the moment which I think is the cause of the Travis failure...

dra27 commented:
Rebased onto trunk and comments updated to reflect that fact it wasn't included in 4.03.0

mshinwell commented:
I'm a bit sceptical of this approach.  It looks similar to what we used to do at Jane Street, for internal compiler builds, which we have recently superceded by distinguished git branches containing the appropriate compiler plus any patches.  Patches aren't that easy to manage: they may conflict, there may be a particular order in which they need to be applied, they may need to be applied pre- or post- bootstrap, it can be hard to know _for sure_ post-hoc exactly what was applied, etc.  Using the version control system seems more robust.

If there was some way by which every commit (or the majority of commits) were tagged with a "proper" piece of metadata that could be read from the repository during a build (such that it is available in the output of `ocamlc -config`, etc), would that be sufficient?  I'm not sure how feasible that is to implement.

dra27 commented:
I'm not completely clear which bit of it you're sceptical about: the inclusion of the `patches:` line in the output of `ocamlc -config` or the automatic picking up of patches in the `patches/` folder during `configure` (or `make -f Makefile.nt patches` for Windows). Or both?
If it's the applying of patches (which I certainly agree can be volatile very quickly: I'm having great 'fun' balancing myriad backports on [metastack/ocaml-legacy](https://github.com/metastack/ocaml-legacy)!), then that's not the main part of this patch. The principal feature is that the `patches:` line whose content is generated based on files with `.patched` found in the `patches` folder. For the model describe, that would simply mean checking in `patches/*git-branch-name*.patched` on the git branch.
See, for example, [metastack/ocaml-legacy/GPR658.patch](https://github.com/metastack/ocaml-legacy/blob/master/GPR658.patch#L30-L34) - it's up to the user to have applied GPR658.patch before running configure, but it still causes gpr#658 to be displayed in the `patches:` line of `ocamlc -config`

mshinwell commented:
I'm not disagreeing about `ocamlc -config`.  It's automatic picking up of patches that I think is a bad idea.  You point out another reason why in your comment: the user might just forget to apply the patch.  I think it's much more robust to do it via the version control system.  (That could also include a check to determine whether the tree is clean, i.e. everything committed, at the time of building.)

damiendoligez commented:
I agree with @mshinwell. IMO applying the right patches in the right order is a job for the VCS or the package manager, not for the configure script.

OTOH, adding the `patches:` line to `ocamlc -config` is a really good idea because it provides a standard way to communicate between patch authors and users.

dra27 commented:
@mshinwell - slightly at crossed purposes, I think. If the user forgets to apply a manual patch (that is one which automatically places a file with a `.patched` extension into `patches/` and must be applied before running `configure` or `make`) then it won't appear in the `patches:` line of `ocamlc -config` ever. In many respects, having the `.patched` file in the VCS is the **more** brittle way - because that file could be part of a stray merge and so result in a compiler claiming to have a patch which it doesn't.

On the other hand, if you drop `foo.patch` into `patches/` the **only** way you will get a compiler which will claim to have `foo` in amongst its patches is if `foo.patch` has been _successfully_ applied (the make fails if a patch doesn't apply).

However, as I said, I'm far more interested in the `patches:` line in `-config` than how the patches are applied! Given that "applying" the patches via a VCS branch is already possible (you just need to check in `patches/foo.patched` on that branch), do you (both) object to the automatic application of `patches/*.patch` strongly enough to want it removed?

damiendoligez commented:
@dra27 I really think the automatic patching is not worth the added complexity in `configure` and the makefiles.

dra27 commented:
Those two are just a rebase of the existing commits - it was "painful", so just waiting for the CI to comment. I'm in the process of removing the automated patching on my local copy...

dra27 committed:
Remove the automatic patching system

Augment the information displayed by the -config option of ocamlc and
ocamlopt to include details of installed patches. utils/config.ml now
contains a string list patches whose elements are the filenames (without
extension) of patches/*.patched (thus, if a patch is applied, it need
simply touch patches/foo.patched to have foo included in the list of
applied patches).

The intention is to provide a slightly more fine-grained approach than
encoding everything in +add-on in the OCaml version number.

dra27 commented:
Right, I've left this as three commits for the time being - my preference would be to squash the first two commits (using the commit comment of the second).

The "Provide patches for back-porting" still provides the automatic patch system, but I'm not viewing that as a problem - I can regenerate them, I was just avoiding having to test the patches against all those OCamls...

dra27 commented:
I've rebased this on trunk, but that actually merging on 4.04. Any further comments on this, especially as regards squashing the first two commits?

Cherry picking these to 4.04 is simply a matter of ensuring that the ocamlnat stuff isn't pulled into 4.04 by mistake in `Makefile.shared` (I can open a separate GPR for 4.04 if that'd help?)

lefessan commented:
Why not add a line `features:` instead of `patches:` ? For me, it makes more sense to check for a particular additionnal feature than for a patch. In particular, new versions of OCaml could also add features in that list, so that, instead of checking for a version, a program could check for a feature. For example, OCamlPro backported `match... with exception` into its own `4.01` version, so some programs could have just checked for `MatchException` instead of the version.

Also, about the directory approach: I would prefer to have a `features` variable in `Clflags` or `Config`, and each patch would just add a line `Clflags.features := "MyFeature" :: !Clflags.features`. This way, we keep the compositionnality of patches without relying on the File System structure.

lefessan commented:
Also, why is it still targetted to `4.04`, 2 weeks after the freeze ? New features should now target only trunk, while `4.04` should only accept bug fixes.

damiendoligez commented:
I think @lefessan's suggestions are good, especially when they replace large amounts of shell/Makefile code with simple OCaml instructions.

Opinions?

dra27 commented:
"large amounts" in the current iteration is only one line of `Makefile` code...! :wink:

I agree in principle that any ML approach is neater than shell programming, of course, but I'm left wondering here if it's a good idea. Doing it this way would virtually eliminate the ability to do it using patch files because the context lines in whichever file you place the `Clflags.features := ...` lines will break in a "non-fuzzy" manner once several patches start trying to add to it. As far as I can see, you'd end up with patches needing to add `echo "Clflags := .." >> utils/config.mlp"` lines to `configure` which seems considerably more brittle. It also removes the guarantee of the format of items in the list and also their uniqueness, though that's of course fairly trivial to solve. For the sake of one line broken into two in one `Makefile`, I'm not wholesomely convinced of the benefit?

Regarding `features` vs `patches` in the output, I guess for me this is more about patches (see the ocaml-legacy repo) so that name made more sense to me. However, the idea of calling it features has made me think further - perhaps all features should be being implemented this way. Long-term support distros (RHEL, etc.) hate build systems which make assumptions about things being available based on version numbers because that makes back-porting a nightmare. Maybe this should really be a larger feature which would boils down to mechanising the `Changes` file. New features would then either need official names or be referenced by a GPR number. For example, 4.04 would include `gadt-equations`, `pattern-local-opening`, `local-exceptions`, `extension-semis`, `unboxed-attribute`, `gpr#464`, `gpr#475`, `gpr#591`. Essentially, it means every fix should add something to this list, to allow the possibility that someone else may choose to back-port it?

Obviously, at this point the list clearly becomes very large - it would become more sensible to add an extra parameter to ocamlc and ocamlopt at this point. For example, say you want to use the extensible open types added in 4.02.0, at present a configure script would test the version of OCaml. However, OPAM already has that feature back-ported to both 4.00.1 and 4.01.0 which would not be picked up by that test. Instead, this could be:

``` bash
if [ $(ocamlc -config | fgrep -x "feature_probing: true") ] ; then
  HAS_OPEN_TYPES=$(ocamlc -query-feature open-types)
else
  if [[ $(ocamlc -vnum | sed -e "s/\.//g") -ge 4020 ]] ; then
    HAS_OPEN_TYPES=yes
  else
    HAS_OPEN_TYPES=no
  fi
fi
```

For configure-based systems, that could easily be reduced to a macro `AC_CHECK_OCAML_FEATURE(open-types)` where the autoconf macro knows that open-types was introduced in 4.02.0.

There might also be some extra flag `ocamlc -list-features` ... the list would be enormous, so it doesn't feel like it wants to go in -config (or, alternatively, `ocamlc -config -verbose`?)

Further thoughts on this larger feature? I was already thinking about using Git hooks to validate merges, and the CI already tests for alterations to `Changes`...

damiendoligez commented:
> Doing it this way would virtually eliminate the ability to do it using patch files because the context lines in whichever file you place the `Clflags.features := ...` lines will break in a "non-fuzzy" manner once several patches start trying to add to it.

I think the idea was to add them in the file you are patching, as close as possible to the changed code, so they would not end up interfering with each other.

About the `-query-feature` option, I'd just go with `-list-features` and I don't think it would become so large that `fgrep` would start being a bottleneck.

But let's take one more step back: do we want to get so serious about backporting that we want to build everything around it?

This whole "testing for features" thing is neat, but the feature set has a lot of structure: each version of OCaml is mostly adding to it, and back-ported features are quite rare, so it makes a lot of sense to only list features that were added to the vanilla version.

xavierleroy commented:
Three years later, any decision about this proposal?
